code/game/machinery/autolathe/nanoforge.dm

```diff
@@ -0,0 +1,100 @@
+/obj/machinery/autolathe/nanoforge
+	name = "matter nanoforge"
+	desc = "It consumes items and produces compressed matter."
+	icon_state = "nanoforge"
+	icon = 'icons/obj/machines/autolathe.dmi'
+	use_oddities = TRUE
+	is_nanoforge = TRUE
+	circuit = /obj/item/weapon/electronics/circuitboard/nanoforge
+	var/list/nanoforge_designs = list()
+	var/list/tags_to_spawn = list(SPAWN_DESING)
+	var/list/nano_disks = list()
+
+	var/static/list/matter_to_compressed = list(
+	MATERIAL_IRON = 0.6,
+	MATERIAL_GLASS = 0.4,
+	MATERIAL_STEEL = 0.3,
+	MATERIAL_PLASMAGLASS = 0.6,
+	MATERIAL_DIAMOND = 0.7,
+	MATERIAL_PLASMA = 0.5,
+	MATERIAL_GOLD = 0.6,
+	MATERIAL_URANIUM = 0.7,
+	MATERIAL_SILVER = 0.6,
+	MATERIAL_PLASTEEL = 0.6,
+	MATERIAL_PLASTIC = 0.4,
+	MATERIAL_TRITIUM = 0.7,
+	MATERIAL_PLATINUM = 0.7,
+	MATERIAL_MHYDROGEN = 0.8,
+	MATERIAL_WOOD = 0.1,
+	MATERIAL_CLOTH = 0.1,
+	MATERIAL_CARDBOARD = 0.1,
+	MATERIAL_RGLASS = 0.4,
+	MATERIAL_LEATHER = 0.1,
+	MATERIAL_TITANIUM = 0.6)
+
+/obj/machinery/autolathe/nanoforge/proc/compress_matter(mob/user)
+	if(!inspiration || !inspiration.perk)
+		to_chat(user, SPAN_WARNING("Catalyst not found."))
+		return
+	var/compressed_amt
+	for(var/mat in stored_material)
+		compressed_amt += stored_material[mat] * matter_to_compressed[mat]
+	if(compressed_amt < 1)
+		to_chat(user, SPAN_WARNING("Not enough materials."))
+	else
+		if(user)
+			var/list/options = list("yes", "no")
+			var/option = input(user, "Proceed?", "Compressing matter will use all of the stored materials", null) as null|anything in options
```

Comment:
This point can be used to duplicate material. Recalculate `compressed_amt` and check it again after user input.

Reply:
Same problem as above.